False positive for `G201` and `G202`?

We may have a situation where we have a callback function defined to handle exception, e.g. this generic catch-all when using a resolver from `aws-lambda-powertools`:

```python
from __future__ import annotations

from typing import Any

from aws_lambda_powertools.event_handler import ALBResolver, Response
from aws_lambda_powertools.logging import Logger
from aws_lambda_powertools.utilities.typing import LambdaContext

app = ALBResolver()
logger = Logger()


@app.get("/")
def handle_default() -> Response:
    raise Exception  # Something went wrong...


@logger.inject_lambda_context
def lambda_handler(event: dict[str, Any], context: LambdaContext) -> dict[str, Any]:
    return app.resolve(event, context)


@app.exception_handler(Exception)
def handle_exception(exc: Exception) -> Response:
    event = app.current_event
    extra = {"path": event.path, "method": event.http_method, "status": 500}
    message = "An unexpected error occurred."
    logger.error(message, exc_info=exc, extra=extra)
    return Response(
        status_code=500,
        payload={"error": message},
        content_type="application/json",
    )
```

I get the following:

```console
❯ ruff --version
ruff 0.0.237

❯ ruff --select G bug.py 
bug.py:28:12: G201 Logging `.exception(...)` should be used instead of `.error(..., exc_info=True)`
Found 1 error.
```

For context, `Logger.exception()` is defined as:

```python
class Logger(Filterer):
    ...

    def exception(self, msg, *args, exc_info=True, **kwargs):
        """
        Convenience method for logging an ERROR with exception information.
        """
        self.error(msg, *args, exc_info=exc_info, **kwargs)

    ...
```

There are two issues here:

1. I'm not using `exc_info=True`, I'm explicitly passing the exception object
2. The guidance to use `.exception(...)` really only applies in an `except:` block
   - This is because passing `exc_info=True` calls `sys.exc_info()` under the hood
   - And `sys.exc_info()` cannot be guaranteed to have an exception outside of `except:`
      ```console
      ❯ python -c 'import sys; print(sys.exc_info())'
      (None, None, None)
      ```

Given the definition of `.exception(...)`, it looks as though we ought to be able to bypass `G201` by passing `exc_info=exc` to `.exception(...)` instead of `.error(...)`, but that gives the following:

```console
❯ ruff --select G bug.py 
bug.py:28:31: G202 Logging statement has redundant `exc_info`
Found 1 error.
```

Which is also incorrect... It's only redundant if it is `exc_info=True`, the default value.

So I believe the following should change:
- Only flag up `G201` in an `except:` block and only when `.error(...)` is passed `exc_info=True` or `exc_info=sys.exc_info()`
- Only flag up `G202` for `.exception(..., exc_info=True)` or `.exception(..., exc_info=sys.exc_info())`
- Consider whether `.exception(..., exc_info=exc)` should be forced back to `.error(..., exc_info=exc)` with a new rule
  - _The only benefit of `.exception(...)` is to pass `exc_info=True` automatically. It's syntactic sugar._

--- a/resources/test/fixtures/flake8_logging_format/G201.py
+++ b/resources/test/fixtures/flake8_logging_format/G201.py
@@ -1,3 +1,21 @@
 import logging
+import sys
 
-logging.error('Hello World', exc_info=True)
+# G201
+try:
+    pass
+except:
+    logging.error("Hello World", exc_info=True)
+
+try:
+    pass
+except:
+    logging.error("Hello World", exc_info=sys.exc_info())
+
+# OK
+try:
+    pass
+except:
+    logging.error("Hello World", exc_info=False)
+
+logging.error("Hello World", exc_info=sys.exc_info())

--- a/resources/test/fixtures/flake8_logging_format/G202.py
+++ b/resources/test/fixtures/flake8_logging_format/G202.py
@@ -1,3 +1,21 @@
 import logging
+import sys
 
-logging.exception('Hello World', exc_info=True)
+# G202
+try:
+    pass
+except:
+    logging.exception("Hello World", exc_info=True)
+
+try:
+    pass
+except:
+    logging.exception("Hello World", exc_info=sys.exc_info())
+
+# OK
+try:
+    pass
+except:
+    logging.exception("Hello World", exc_info=False)
+
+logging.exception("Hello World", exc_info=True)

--- a/src/checkers/ast.rs
+++ b/src/checkers/ast.rs
@@ -93,6 +93,7 @@ pub struct Checker<'a> {
     in_type_definition: bool,
     in_deferred_string_type_definition: bool,
     in_deferred_type_definition: bool,
+    in_exception_handler: bool,
     in_literal: bool,
     in_subscript: bool,
     in_type_checking_block: bool,
@@ -153,6 +154,7 @@ impl<'a> Checker<'a> {
             in_type_definition: false,
             in_deferred_string_type_definition: false,
             in_deferred_type_definition: false,
+            in_exception_handler: false,
             in_literal: false,
             in_subscript: false,
             in_type_checking_block: false,
@@ -1714,6 +1716,7 @@ where
         }
 
         // Recurse.
+        let prev_in_exception_handler = self.in_exception_handler;
         let prev_visible_scope = self.visible_scope.clone();
         match &stmt.node {
             StmtKind::FunctionDef {
@@ -1865,9 +1868,13 @@ where
                 }
                 self.visit_body(body);
                 self.except_handlers.pop();
+
+                self.in_exception_handler = true;
                 for excepthandler in handlers {
                     self.visit_excepthandler(excepthandler);
                 }
+                self.in_exception_handler = prev_in_exception_handler;
+
                 self.visit_body(orelse);
                 self.visit_body(finalbody);
             }
@@ -3779,6 +3786,10 @@ impl<'a> Checker<'a> {
             .iter()
             .rev()
             .map(|index| &self.scopes[*index])
+    }
+
+    pub fn in_exception_handler(&self) -> bool {
+        self.in_exception_handler
     }
 
     pub fn execution_context(&self) -> ExecutionContext {

--- a/src/rules/flake8_logging_format/rules.rs
+++ b/src/rules/flake8_logging_format/rules.rs
@@ -205,7 +205,28 @@ pub fn logging_call(checker: &mut Checker, func: &Expr, args: &[Expr], keywords:
                     .rules
                     .enabled(&Rule::LoggingRedundantExcInfo)
             {
+                if !checker.in_exception_handler() {
+                    return;
+                }
                 if let Some(exc_info) = find_keyword(keywords, "exc_info") {
+                    // If `exc_info` is `True` or `sys.exc_info()`, it's redundant; but otherwise,
+                    // return.
+                    if !(matches!(
+                        exc_info.node.value.node,
+                        ExprKind::Constant {
+                            value: Constant::Bool(true),
+                            ..
+                        }
+                    ) || if let ExprKind::Call { func, .. } = &exc_info.node.value.node {
+                        checker.resolve_call_path(func).map_or(false, |call_path| {
+                            call_path.as_slice() == ["sys", "exc_info"]
+                        })
+                    } else {
+                        false
+                    }) {
+                        return;
+                    }
+
                     match logging_level {
                         LoggingLevel::Error => {
                             if checker.settings.rules.enabled(&Rule::LoggingExcInfo) {

